server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java

```diff
@@ -161,6 +164,26 @@ public void setLastAccessTime() {
         lastAccessTime = System.nanoTime();
     }
 
+    /**
+     * Set flag to signal to {@link org.elasticsearch.indices.recovery.RecoveriesCollection.RecoveryMonitor} that it must not cancel this
+     * recovery temporarily. This is used by the primary relocation mechanism to avoid recovery failure in case a long running relocation
+     * condition was added to the shard via {@link IndexShard#createRelocationDependency()}.
+     *
+     * @return releasable that once closed will re-enable liveness checks by the recovery monitor
+     */
+    public Releasable disableRecoveryMonitor() {
+        assert recoveryMonitorEnabled : "recovery monitor already disabled";
+        recoveryMonitorEnabled = false;
+        return () -> {
+            setLastAccessTime();
+            recoveryMonitorEnabled = true;
```

Comment:
This is a little low-tech relative to the tricky ref-counting in the `IndexShard`.  I figured this was ok here since the hand-off request only comes in once (at least judging by the assertions we have in `IndexShard`) while the other API has a more "feel" to it and there are no hard guarantees on the index shard state listener only being invoked once (though the "loaded" flag on the directory effectively guarantees we only add one condition for now) and it wasn't that much extra effort since the API was supposed to be non-blocking anyway.